spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala

```diff
@@ -209,52 +276,40 @@ class DLModel[@specialized(Float, Double) T: ClassTag](
 
   /**
    * Perform a prediction on featureCol, and write result to the predictionCol.
-   * @param featureData featureData in the format of Seq
-   * @return output DataFrame
    */
-  protected override def internalTransform(
-      featureData: RDD[Seq[AnyVal]], dataset: DataFrame): DataFrame = {
-
-    model.evaluate()
-    val modelBroadCast = ModelBroadcast[T]().broadcast(featureData.sparkContext, model)
-    val predictRdd = featureData.map { f =>
-      // convert feature data type to the same type with model
-      f.head match {
-        case dd: Double => f.asInstanceOf[Seq[Double]].map(ev.fromType(_))
-        case ff: Float => f.asInstanceOf[Seq[Float]].map(ev.fromType(_))
-      }
-    }.mapPartitions { feature =>
-      val localModel = modelBroadCast.value()
-      val tensorBuffer = Tensor[T](Array($(batchSize)) ++ featureSize)
-      val batches = feature.grouped($(batchSize))
-      batches.flatMap { batch =>
-        var i = 1
-        // Notice: if the last batch is smaller than the batchSize, we still continue
-        // to use this tensorBuffer, but only add the meaningful parts to the result Array.
-        batch.foreach { row =>
-          tensorBuffer.select(1, i).copy(Tensor(Storage(row.toArray)))
-          i += 1
-        }
-        val output = localModel.forward(tensorBuffer).toTensor[T]
-        val predict = batchOutputToPrediction(output)
-        predict.take(batch.length)
+  protected override def internalTransform(dataFrame: DataFrame): DataFrame = {
+    val featureType = dataFrame.schema($(featuresCol)).dataType
+    val featureColIndex = dataFrame.schema.fieldIndex($(featuresCol))
+    val featureFunc = getConvertFunc(featureType)
+
+    val samplesRDD = dataFrame.rdd.map { row =>
+      val features = featureFunc(row, featureColIndex)
+      val featureBuffer = features.head match {
+        case dd: Double => features.asInstanceOf[Seq[Double]].map(ev.fromType(_))
+        case ff: Float => features.asInstanceOf[Seq[Float]].map(ev.fromType(_))
       }
+      Sample(Tensor(featureBuffer.toArray, featureSize), Tensor[T](1))
     }
 
-    val resultRDD = dataset.rdd.zip(predictRdd).map { case (row, predict) =>
-      Row.fromSeq(row.toSeq ++ Seq(predict))
-    }
-    val resultSchema = transformSchema(dataset.schema)
-    dataset.sqlContext.createDataFrame(resultRDD, resultSchema)
+    val prediction = model.asInstanceOf[Module[T]]
+      .predict(samplesRDD, $(batchSize))
+      .map { t => outputToPrediction(t.asInstanceOf[Tensor[T]]) }
+
+    val resultRDD = dataFrame.rdd.zipWithIndex().map(_.swap)
```

Comment:
A potential issue is that dataFrame.rdd may need to be computed twice if it's not cached. One way to address that is to do something like https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala#L93